components/collator/src/comparison.rs

```diff
@@ -1543,4 +1644,660 @@ impl CollatorBorrowed<'_> {
 
         Ordering::Equal
     }
+
+    fn sort_key_levels(&self) -> u8 {
+        #[allow(clippy::indexing_slicing)]
+        let mut levels = LEVEL_MASKS[self.options.strength() as usize];
+        if self.options.case_level() {
+            levels |= CASE_LEVEL_FLAG;
+        }
+        levels
+    }
+
+    /// Write the sort key bytes up to the collator's strength.
+    ///
+    /// For identical strength, the UTF-8 NFD normalization is appended for breaking ties.
+    /// The key ends with a `TERMINATOR_BYTE`.
+    pub fn write_sort_key<S>(&self, s: &str, sink: &mut S)
+    where
+        S: Write,
+    {
+        self.write_sort_key_up_to_quaternary(s.chars(), sink, |_| true);
+
+        if self.options.strength() == Strength::Identical {
+            let nfd =
+                DecomposingNormalizerBorrowed::new_with_data(self.decompositions, self.tables);
+            let s = nfd.normalize(s);
+            sink.write_byte(LEVEL_SEPARATOR_BYTE);
+            sink.write(s.as_bytes());
+        }
+
+        sink.write(&[TERMINATOR_BYTE]);
+    }
+
+    /// Write the sort key bytes up to the collator's strength.
+    ///
+    /// Optionally write the case level.  Stop writing levels when `need_to_write` returns
+    /// `false`.  Separate levels with the `LEVEL_SEPARATOR_BYTE`, but do not write a
+    /// `TERMINATOR_BYTE`.
+    pub fn write_sort_key_up_to_quaternary<I, S, W>(&self, iter: I, sink: &mut S, need_to_write: W)
+    where
+        I: Iterator<Item = char>,
+        S: Write,
+        W: Fn(Level) -> bool,
+    {
+        // The following code started as a port from ICU4C which carried the
+        // following notice:
+        //
+        // // © 2016 and later: Unicode, Inc. and others.
+        // // License & terms of use: http://www.unicode.org/copyright.html
+        // /*
+        // *******************************************************************************
+        // * Copyright (C) 2012-2015, International Business Machines
```

Comment:
@srl295 Given that we have the IBM copyright notice with a broader year range already in https://github.com/unicode-org/icu4x/blob/09d75a966aaeb2a302f195ce4a91bfa8a8eb5d1b/LICENSE#L45 , do we need this license block to state the license header of the relevant ICU4C file, or can this whole block comment be shortened to something like:

"The algorithm comes from `CollationKeys::writeSortKeyUpToQuaternary` in ICU4C."